### ISSUE TITLE
[FEATURE] Bind shell agents and Tor support

### ISSUE THREAD

**Issue by Strazzom** [EmpireProject/Empire#981]
**Bind shell agents:**

TL;DR: I want bind agents, and I might implement Tor functionality myself for hidden service access. Benefits: meek domain fronting, no CNC, simple to write.

Bind shells are often discounted because they are generally less useful than reverse shells. Usually (unless you have local LAN access) a bind shell is not accessible from the public internet (unless you have modified the firewall/router to allow the port). However, there are circumstances where they are useful.

Take the [APT29 hidden service backdoor](https://www.fireeye.com/blog/threat-research/2017/03/apt29_domain_frontin.html) for example. They basically wrote a program to start a Tor hidden service on the host, and connected to RDP through that hidden service. In an ideal world, they would have been using [hidden service authentication](https://www.whonix.org/wiki/Hidden_Services#Hidden_Service_Authentication) for security, that way it would not have been publicly accessible. Ignoring the fact that RDP must have been ridiculously slow over a Tor connection, they were restricted to using built in Windows tools to perform exfiltration. Empire is uniquely tailored for espion-**cough** "penetration testing", so it would be an asset to be able to use it over Tor in some capacity. Tor is useful, because it is basically impossible to track down the C2, and it has built in domain fronting (meek bridges).

The other benefit of the APT29 approach is their lack of a C2 in the first place. In their case, the client rather than the server is running the hidden service. Because of the way hidden services work, they punch through the firewall and enable someone to access ports running on a host without needing to open ports on the NAT/router/firewall. It basically behaves like the port had been forwarded with a metasploit listener (bind shell) running on it. So, because of that lack of a central server, there is no single point of failure that can be taken down, and you also have the added benefit of not needing to pay for that server (VPS) in the first place.

Regardless of what approach is being used, we can't deny the benefit of having Tor integrated with Empire, either with bind listeners or a regular reverse tcp connection to a hidden service CNC. I have outlined ways to implement both approaches below:

In both cases, a Tor client running on the client machine would be necessary. This is fairly easy to do, and it should be relatively simple to implement this as an Empire module/listener (not sure what category it would go in, seeing as this is directly tied in with listener functionality). I am willing to do this myself, although I am at best a mediocre programmer.

**Reverse shell over Tor:**

To do this, there are two prerequisite features that must be implemented: 1) Proper socks4/5 proxy support, and 2) [Remote DNS.](https://trac.torproject.org/projects/tor/wiki/doc/FirefoxRemoteDNS)

Long story short: agent should connect through the Tor proxy running on localhost:9050, and should resolve DNS queries through the Tor DNS port (so that it can resolve .onion addresses). It should connect back to a C2 running as a hidden service.

This is relatively complex to implement, and I personally am not partial to it. Both prerequisite features look like a hassle. However, there may be people who have a use for it.

**Bind shell over Tor:**

This is simple. All one would need to do is have the server connect to an agent listening on a local port on the client, similar to metasploit bind shells. The Tor client would start a hidden service with the port the agent is listening on. The Empire server connects to the client through the hidden service and issues commands. For authentication, I would personally use hidden service auth, but it could also be built into the bind agent.

Obviously, there are issues with the server resolving the onion address of the client, however this can be solved more easily on the server side (Whonix gateways, forcing connection over Tor with Privoxy, etc.)

**Bind shells generally:**

Regardless of whether it is used for Tor or just by itself, bind shells are a feature that desperately need to be in Empire, because they create new ways to extend existing functionality. It could be as simple as having a bind shell listener, with options like:

`OpenPort: open port in system firewall to allow LAN access to agent.` (not necessary to open port with hidden services, but useful for local LAN access.)
`PortAuthentication: enable key/cert based authentication for agent.` (should be simple to implement, SSL/key exchange is already there. Only useful for securing opened firewall port when you have LAN access.)

It could also communicate over http, much like the regular reverse shell listener.

**Comment by xorrior** [comment]
@Strazzom This sounds really cool. Unfortunately, I don't have the cycles to implement a major feature request like this. However, feel free to give it a shot and submit a PR. We can go from there. Thanks
